mimxrt: Fix SD card deadlock and implement proper timeout handling.

andrewleech opened the pull request:
## Summary

Fixes two bugs in the mimxrt SD card driver: 
* Deadlock when no card is present and ineffective timeout in transfer operations. 
* The timeout fix uses the non-blocking USDHC API with interrupt-driven callbacks, following the pattern used by sdio.c.

### Issue 1: Deadlock on missing card

On a board with no SD Card Detect pin, when attempting to mount an SD card filesystem in `_boot.py` with no card inserted the system would deadlock in `sdcard_power_off()`. The function unconditionally sent GO_IDLE_STATE command even when the card was never successfully initialized. Since the USDHC driver's `USDHC_WaitCommandDone()` has an infinite loop waiting for command completion, this caused a permanent hang when no card was present to respond.

This deadlock occurs in mimxrt/machine_sdcard.c during the initial ioctl / `case MP_BLOCKDEV_IOCTL_INIT` which tries to `sdcard_power_on()` and if that fails, runs `sdcard_power_off()` which calls down to `sdcard_cmd_go_idle_state()` which deadlocks when there's no card available.

### Issue 2: Ineffective timeout in blocking transfers

The `sdcard_transfer_blocking()` function had a timeout loop that checked if the card was busy before calling `USDHC_TransferBlocking()`. However, this timeout was ineffective because `USDHC_TransferBlocking()` contains infinite loops in `USDHC_WaitCommandDone()` and `USDHC_TransferDataBlocking()` with no timeout mechanism. Once inside the SDK blocking call, the timeout loop couldn't abort the operation.

This PR uses the non-blocking USDHC API from the SDk instead, matching the pattern already used in `ports/mimxrt/sdio.c`.

## Testing

Tested on PHYBOARD-RT1170 with:

- No SD card inserted (previously deadlocked, now returns error gracefully)
- SD card insertion/removal during operation
- Normal read/write operations with SD card present
- All operations now have working timeouts

robert-hh commented:
I tried the PR with a little test program at a Teensy 4.1, which does not have a card detect pin, and a MIMXRT1020_EVK with a card detect pin. Both fail the test at line 26 and not data is written. The master branch code passes. Test script:
```py
import os
import random
from machine import SDCard
from time import ticks_us
import gc

sdcard = SDCard(1)
def run(size_mult=1):
    gc.collect()
    size = (411 + random.randint(1, 1 << 16)) & 0xfffffe00
    data = bytearray(size)
    print("size = ", size)

    os.chdir("/sdcard")
    print(os.getcwd())
    print(os.listdir())

    ticks_start_us = 0
    ticks_end_us = 0
    ticks_start_us = ticks_us()
    mode = "wb"
    f = open("test_file.txt" , mode)
    for _ in range(size_mult):
        print("loop", _, len(data))
        f.write(data)
    f.close()
    ticks_end_us = ticks_us()

    data_size_mb = (len(data) * size_mult)/(1024.0*1024.0)
    duration = (ticks_end_us - ticks_start_us)/1000000.0
    data_rate = data_size_mb / duration

    print()
    print("Block Length    :", len(data))
    print("Total duration  : {:.1f} s".format(duration))
    print("Transmitted data: {:.1f} Mb".format(data_size_mb))
    print("Data Rate       : {:.1f} MB/s".format(data_rate))

    return data_rate

run()
```

robert-hh commented:
The last version of your PR still fails. Until now, I used only the Teensy 4.1 for the test. I tried it with a MIMXRT1020EVK, Arch MIX and MIMXRT1170_EVK. The test **passes** at these boards. The major difference between especially the Teensy and the other boards is SDRAM. Teensy does not have SDRAM, the other boards have. So I disabled SDRAM at the Arch Mix (mpconfigboard.mk, flag MICROPY_HW_SDRAM_AVAIL). Then, the Arch Mix board shows the same error.

andrewleech commented:
Thanks @robert-hh I had a few more attempts at getting it working with the non blocking SDK API but just could not figure out how to prevent the corruption. 

I did briefly think we could keep both implementations, `#ifdef` sdram enabled, but that does complicate things a fair bit. 

The only reason I can think of to push for the non blocking API would be that while it's waiting it can use mp_event_wait() which lets scrolled tasks run. Probably not a big deal, but a tangible benefit. What do you think?

pi-anl committed:
mimxrt/sdcard: Improve robustness of sdcard driver.

Replace interrupt-based transfer with polling (USDHC_TransferBlocking)
and add timeouts throughout to prevent deadlocks. Add ERR050396
workaround to disable cacheable AXI transactions for USDHC, and
explicit D-cache management around DMA transfers. Add sdcard_wait_ready
polling after writes to ensure programming completes. Guard
sdcard_power_off against sending commands when no card is present.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

projectgus commented:
@andrewleech #19304 has just merged, so unfortunately you have a bit of a messy rebase on your plate now (sorry). The good news is that I think some of this PR should become simpler (and if we're lucky the corruption issue will go away, but have to wait and see...)

andrewleech commented:
No worries thanks @projectgus - I want to pivot back to the interrupt based method anyway based on your work inn this space and see it out works now! I just need to get my mimx Feb hardware plugged together again at home... I'm spending too much time in the office lately ;-)
